source/manual/conventions-in-rails-applications.html.md

```diff
@@ -0,0 +1,301 @@
+where this needs to be associated with a database [ActiveStorage][] should be
+used.
+
+[ActiveRecord]: https://guides.rubyonrails.org/active_record_basics.html
+[db/seeds.rb]: https://github.com/alphagov/content-publisher/blob/master/db/seeds.rb
+[key-value datastore]: https://en.wikipedia.org/wiki/Key-value_database
+[Redis]: https://redis.io/
+[Amazon S3]: https://aws.amazon.com/s3/
+[ActiveStorage]: https://guides.rubyonrails.org/active_storage_overview.html
+
+## Background job processing
+
+The preferred approach for background job processing is to use [ActiveJob][]
+with [Sidekiq][] with the [govuk_sidekiq][] gem providing configuration.
+ActiveJob is preferred due to it's closer integration of Rails components
+(such as [ActionMailer][]).
+
+Scheduled background jobs for applications should also make use of Sidekiq,
+where [sidekiq-scheduler][] is the conventional choice to achieve this.
+
+[Sidekiq]: https://sidekiq.org
+[ActiveJob]: https://guides.rubyonrails.org/active_job_basics.html
+[ActionMailer]: https://guides.rubyonrails.org/action_mailer_basics.html
+[sidekiq-scheduler]: https://github.com/moove-it/sidekiq-scheduler
+
+## JavaScript package management
+
+As per Rails conventions GOV.UK applications should use
+[Yarn](https://yarnpkg.com) for JavaScript packages. This is because it
+integrates better with Rails than [NPM](https://www.npmjs.com/).
+
+## Frontend assets
+
+GOV.UK applications use the Rails [asset pipeline][] for building assets. We
+aspire to migrate to [webpacker][] as the conventional approach when we have
+resolved [blocking issues][].
+
+[asset pipeline]: https://guides.rubyonrails.org/asset_pipeline.html
+[webpacker]: https://github.com/rails/webpacker
+[blocking issues]: https://github.com/alphagov/govuk_publishing_components/issues/505
+
+## Sending emails
+
+[GOV.UK Notify][] is the preferred approach for sending email. The
+[mail-notify][] gem is conventionally used to integrate Notify with
+[ActionMailer][]. There is a shared GOV.UK Publishing Notify account that can
+be used across GOV.UK publishing/admin applications.
+
+[GOV.UK Notify]: https://www.notifications.service.gov.uk/
+[mail-notify]: https://github.com/dxw/mail-notify
+
+## Configure Rails as `api_only` for API projects
+
+For applications that do not serve requests to a web browser you should
+configure the Rails application to be [API only][].
+
+[API only]: https://guides.rubyonrails.org/api_app.html
+
+## Organising your code
+
+Organising code in a Rails application is frequently a source of debate and
+confusion, [particularly as to what code belongs in the `/app` directory and
+what belongs in the `/lib` directory][app-lib]. We aim to follow what appears
+to be the prevalent convention where:
+
+- `/app` is used for:
+  - the default Rails components (such as controllers, models, views, helpers, jobs)
+  - commonly used classes in the application that have a common purpose and
```

Comment:
Sure, can certainly try. Any suggestions?

Reply:
I think there are quite a few good examples, although they tend to be "presenters", which perhaps isn't something we want to encourage. Maybe "policies" or even "interactors"?

Bad example: https://github.com/alphagov/specialist-publisher/tree/d6b489d66641de4cd4c557e08ac2c7d70deaceab/app/reporters

Reply:
What's wrong with presenters?

Reply:
As for examples, smart-answers have an _interesting_ layout: https://github.com/alphagov/smart-answers.

Reply:
I've updated the doc with a couple of examples.

@leenagupte Rather than presenters as a whole being bad, our GOV.UK use of them is inconsistent which makes them somewhat meaningless as a concept. The main problem I'd suggest we have is that what we call a "presenter" can mean a variety of things, which isn't the same for say controllers, views, models, etc and can often reveal smells. Our use that is closest to the [Model-View-Presenter](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93presenter) pattern (and thus, arguably least surprising) is perhaps [GOV.UK Publishing Components](https://github.com/alphagov/govuk_publishing_components/blob/4d4119e0e3424ceefbfb906416ca51d5e3466a03/lib/govuk_publishing_components/presenters/attachment.rb) or [government-frontend](https://github.com/alphagov/government-frontend/tree/e2371b541a7248600099250aed565dfc27233c15/app/presenters) where they act as a model for presentation. However, in other apps we use them for serialisation ([Publishing API](https://github.com/alphagov/publishing-api/blob/eccb8548b286a69adfb83c91f8c74603273a956b/app/presenters/edition_presenter.rb), [Email Alert API](https://github.com/alphagov/email-alert-api/blob/c6fe159dc943deef23d155095996868166624447/app/presenters/content_change_presenter.rb)) and this seems to leave a bit more of a smell as they don't have anything in common with the other type of presenters and are often named after a particular model when they're actually serialising for a particular context.

Reply:
OK, so it's still ok to use Presenters as long as it's for their intended purpose of reading data to be rendered in a view?